core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -494,10 +494,16 @@ private RexNode simplifyNot(RexCall call, RexUnknownAs unknownAs) {
     return call;
   }
 
-  private RexNode simplifyIs(RexCall call) {
+  private RexNode simplifyIs(RexCall call, RexUnknownAs unknownAs) {
     final SqlKind kind = call.getKind();
-    final RexNode a = call.getOperands().get(0);
-
+    final RexNode argument = call.getOperands().get(0);
+    if (kind == SqlKind.IS_TRUE && unknownAs == RexUnknownAs.FALSE) {
+      return simplify(argument, unknownAs);
+    }
+    if (kind == SqlKind.IS_NOT_FALSE && unknownAs == RexUnknownAs.TRUE) {
+      return simplify(argument, unknownAs);
+    }
+    final RexNode a = simplify(call.getOperands().get(0), RexUnknownAs.UNKNOWN);
```

Comment:
This does look like `O(N^2)`.
Can you please avoid simplifying the same nodes again and again?

Reply:
I think this shouldn't be `N^2` ; because it actually drops the "IS X" part and continues with the recursion on the argument.
I've deliberitely placed the existing `simplify` call after the checks of redundancy.

Thank you for taking a look!

Reply:
I guess `simplifyIs` is used in simplifyAnd2 as well, so a generic `simplify` looks very odd inside `simplifyIs`

Reply:
ugh...it seems `simplifyIs2` is also invoking `simplify` - I think recursion should happen as the first thing when entering sub-simplification methods

Reply:
@kgyrtkirk , I agree: https://issues.apache.org/jira/browse/CALCITE-2449 :)